vtk-m@2.1.0 +rocm w/ ROCm 6.1 external: g++: error: language hip not recognized

### Steps to reproduce the issue

`vtk-m@2.1.0 +rocm` build fails using:
* spack@develop (641075539c0a888d4378bd7dfdb19a872c5c42de from `Mon Jun 3 14:37:51 2024 -0700`)
* ROCm 6.1.1 installed via apt, externally defined via `packages:` config
* Ubuntu 22.04, X86_64

We are hitting this as we try to build `ecp-data-vis-sdk +rocm +vtkm ...` with ROCm 6.

We don't see this error using an external install of ROCm 5.7.

Spack environment:  [spack.yaml.txt](https://github.com/user-attachments/files/15917201/spack.yaml.txt)

<details><summary>Concretization</summary>

<pre>
 -   vtk-m@2.1.0%gcc@11.4.0~64bitids~cuda+doubleprecision+examples~fpic~ipo+kokkos~logging+mpi~openmp+rendering+rocm+shared~tbb~testlib amdgpu_target=gfx908 build_system=cmake build_type=Release generator=make patches=64177d0 arch=linux-ubuntu22.04-x86_64_v3
[+]      ^cmake@3.29.4%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64_v3
[+]          ^curl@8.7.1%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64_v3
[+]              ^nghttp2@1.62.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^openssl@3.3.0%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]              ^perl@5.38.2%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]          ^ncurses@6.5%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-x86_64_v3
[+]          ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]      ^hip@6.1.1%gcc@11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,6bbe83b arch=linux-ubuntu22.04-x86_64_v3
[e]      ^hsa-rocr-dev@6.1.1%gcc@11.4.0~asan+image+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[+]      ^kokkos@4.3.01%gcc@11.4.0~aggressive_vectorization~compiler_warnings~cuda~debug~debug_bounds_check~debug_dualview_modify_check~deprecated_code~examples~hpx~hpx_async_dispatch~hwloc~ipo~memkind~numactl~openmp~openmptarget~pic+rocm+serial+shared~sycl~tests~threads~tuning~wrapper amdgpu_target=gfx908 build_system=cmake build_type=Release cxxstd=17 generator=make intel_gpu_arch=none arch=linux-ubuntu22.04-x86_64_v3
[e]      ^llvm-amdgpu@6.1.1%gcc@11.4.0~link_llvm_dylib~llvm_dylib+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=b4774ca arch=linux-ubuntu22.04-x86_64_v3
[+]      ^mpich@4.2.1%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath~xpmem build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=default arch=linux-ubuntu22.04-x86_64_v3
[+]          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-x86_64_v3
[+]          ^libfabric@1.21.0%gcc@11.4.0~debug~kdreg~uring build_system=autotools fabrics=sockets,tcp,udp arch=linux-ubuntu22.04-x86_64_v3
[+]          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^util-macros@1.20.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]          ^pkgconf@2.2.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]          ^yaksa@0.3%gcc@11.4.0~cuda~rocm build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^autoconf@2.72%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^python@3.11.9%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^expat@2.6.2%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                      ^libbsd@0.12.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                          ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^gettext@0.22.5%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                      ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-x86_64_v3
[+]                          ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-x86_64_v3
[+]                          ^zstd@1.5.6%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^libffi@3.4.6%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^util-linux-uuid@2.40.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[e]      ^rocthrust@6.1.1%gcc@11.4.0 amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
</pre>
</details>

Error:
```
$> spack -e . install
...
==> Installing vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j [21/21]
==> No binary for vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j found: installing from source
==> Fetching https://gitlab.kitware.com/vtk/vtk-m/-/archive/v2.1.0/vtk-m-v2.1.0.tar.gz
==> Applied patch /spack/var/spack/repos/builtin/packages/vtk-m/mr3160-rocthrust-fix.patch
==> vtk-m: Executing phase: 'cmake'
==> vtk-m: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j32'

80 errors found in build log:
     155    [  6%] Building CXX object vtkm/cont/CMakeFiles/vtkm_cont.dir/ColorTablePresets.cxx.o
     156    cd /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/cont && /spack/lib/spack/env/gcc/g++ -DUSE_PROF_API=1 -DVTKMDIY_MPI_AS_LIB
             -DVTKMDIY_NO_THREADS -D__HIP_PLATFORM_AMD__=1 -Dvtkm_cont_EXPORTS -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/optionp
            arser -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/diy -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6o
            mtamj43jaztizt6j/spack-src/vtkm/thirdparty/lcl/vtkmlcl -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src -I/tmp/root/spack-stage/spack-stage
            -vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thir
            dparty/diy/vtkmdiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mp
            i -isystem /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/kokkos-4.3.01-3yvzdu5latq72vgeppablkhcqtbyg5hq/include -isystem /opt/rocm-6.1.1/include -O3 -DNDEBUG -std=c++17 -f
            PIC -fvisibility=hidden -Wall -Wcast-align -Wextra -Wpointer-arith -Wformat -Wformat-security -Wshadow -Wunused -fno-common -Wno-unused-function -Wchar-subscripts -Wfloat-conversion
             -Wodr -ffunction-sections -x hip -MD -MT vtkm/cont/CMakeFiles/vtkm_cont.dir/CellSetStructured.cxx.o -MF CMakeFiles/vtkm_cont.dir/CellSetStructured.cxx.o.d -o CMakeFiles/vtkm_cont.d
            ir/CellSetStructured.cxx.o -c /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/cont/CellSetStructured.cxx
     157    [  7%] Building HIP object vtkm/cont/CMakeFiles/vtkm_cont.dir/CellLocatorUniformBins.cxx.o
     158    cd /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/cont && /spack/lib/spack/env/gcc/g++ -DUSE_PROF_API=1 -DVTKMDIY_MPI_AS_LIB
             -DVTKMDIY_NO_THREADS -D__HIP_PLATFORM_AMD__=1 -Dvtkm_cont_EXPORTS -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/optionp
            arser -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/diy -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6o
            mtamj43jaztizt6j/spack-src/vtkm/thirdparty/lcl/vtkmlcl -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src -I/tmp/root/spack-stage/spack-stage
            -vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thir
            dparty/diy/vtkmdiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mp
            i -isystem /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/kokkos-4.3.01-3yvzdu5latq72vgeppablkhcqtbyg5hq/include -isystem /opt/rocm-6.1.1/include -O3 -DNDEBUG -std=c++17 -f
            PIC -fvisibility=hidden -Wall -Wcast-align -Wextra -Wpointer-arith -Wformat -Wformat-security -Wshadow -Wunused -fno-common -Wno-unused-function -Wchar-subscripts -Wfloat-conversion
             -Wodr -ffunction-sections -x hip -MD -MT vtkm/cont/CMakeFiles/vtkm_cont.dir/DataSet.cxx.o -MF CMakeFiles/vtkm_cont.dir/DataSet.cxx.o.d -o CMakeFiles/vtkm_cont.dir/DataSet.cxx.o -c
            /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/cont/DataSet.cxx
     159    [  7%] Building CXX object vtkm/cont/CMakeFiles/vtkm_cont.dir/CellLocatorUniformGrid.cxx.o
     160    cd /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/cont && /spack/lib/spack/env/gcc/g++ -DUSE_PROF_API=1 -DVTKMDIY_MPI_AS_LIB
             -DVTKMDIY_NO_THREADS -D__HIP_PLATFORM_AMD__=1 -Dvtkm_cont_EXPORTS -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/optionp
            arser -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thirdparty/diy -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6o
            mtamj43jaztizt6j/spack-src/vtkm/thirdparty/lcl/vtkmlcl -I/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src -I/tmp/root/spack-stage/spack-stage
            -vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/thir
            dparty/diy/vtkmdiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mp
            i -isystem /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/kokkos-4.3.01-3yvzdu5latq72vgeppablkhcqtbyg5hq/include -isystem /opt/rocm-6.1.1/include -O3 -DNDEBUG -std=c++17 -f
            PIC -fvisibility=hidden -Wall -Wcast-align -Wextra -Wpointer-arith -Wformat -Wformat-security -Wshadow -Wunused -fno-common -Wno-unused-function -Wchar-subscripts -Wfloat-conversion
             -Wodr -ffunction-sections -x hip -MD -MT vtkm/cont/CMakeFiles/vtkm_cont.dir/ColorTablePresets.cxx.o -MF CMakeFiles/vtkm_cont.dir/ColorTablePresets.cxx.o.d -o CMakeFiles/vtkm_cont.d
            ir/ColorTablePresets.cxx.o -c /tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-src/vtkm/cont/ColorTablePresets.cxx
  >> 161    g++: error: language hip not recognized
  >> 162    g++: error: language hip not recognized
...
  >> 261    make[2]: *** [vtkm/cont/CMakeFiles/vtkm_cont.dir/build.make:442: vtkm/cont/CMakeFiles/vtkm_cont.dir/ErrorBadType.cxx.o] Error 1
     262    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao'
     263    [  9%] Built target vtkm_lodepng
     264    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao'
  >> 265    make[1]: *** [CMakeFiles/Makefile2:2820: vtkm/cont/CMakeFiles/vtkm_cont.dir/all] Error 2
     266    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.1.0-xmau3aozad6lrw6omtamj43jaztizt6j/spack-build-xmau3ao'
  >> 267    make: *** [Makefile:159: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
... see aboe
</pre></details>


### Information on your system

* **Spack:** 0.23.0.dev0 (745a1ad003945110d9463383f628b82fa29d9bb8)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/user-attachments/files/15917196/spack-build-env.txt)
[spack-build-out.txt](https://github.com/user-attachments/files/15917197/spack-build-out.txt)

@kmorel @vicentebolea @kwryankrattiger 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin/packages/vtk-m/package.py
@@ -196,6 +196,9 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
                 # vtk-m detectes tbb via TBB_ROOT env var
                 os.environ["TBB_ROOT"] = spec["tbb"].prefix
 
+            if "+kokkos" in spec and "+rocm" in spec and spec.satisfies("^kokkos@4:"):
+                options.append(f"-DCMAKE_CXX_COMPILER:BOOL={spec['hip'].prefix.bin.hipcc}")
+
             # Support for relocatable code
             if "~shared" in spec and "+fpic" in spec:
                 options.append("-DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON")

